dataland-internal-storage/src/main/kotlin/db/migration/V18__CompleteListOfReferencedReports.kt

```diff
@@ -0,0 +1,153 @@
+        val dataSource = jsonObject.getOrJavaNull("dataSource") as JSONObject?
+        return if (dataSource !== null && dataSource.has("fileName") && dataSource.has("fileReference")) {
+            listOf(dataSource)
+        } else {
+            jsonObject.keys().asSequence().mapNotNull { key ->
+                jsonObject.optJSONObject(key)?.let { subJsonObject ->
+                    findAllDataPoints(subJsonObject)
+                }
+            }.flatten().toList()
+        }
+    }
+
+    /**
+     * Gets the referencedReports based on the framework
+     * If the object is null we insert a new object and return this object
+     */
+    private fun getOrInsertReferencedReports(dataset: JSONObject, framework: String): JSONObject {
+        fun getOrInsertJSONObject(jsonObject: JSONObject, key: String): JSONObject {
+            if (jsonObject.optJSONObject(key) == null) {
+                jsonObject.put(key, JSONObject())
+            }
+            return jsonObject.getJSONObject(key)
+        }
+
+        val referencedReports: JSONObject = when (framework) {
+            "eutaxonomy-non-financials" -> {
+                val general = dataset.optJSONObject("general")
+                getOrInsertJSONObject(general, "referencedReports")
+            }
+
+            "eutaxonomy-financials" -> {
+                getOrInsertJSONObject(dataset, "referencedReports")
+            }
+
+            "sfdr" -> {
+                val general = dataset.optJSONObject("general")
+                val generalGeneral = general.optJSONObject("general")
+                getOrInsertJSONObject(generalGeneral, "referencedReports")
+            }
+            else -> {
+                JSONObject()
+            }
+        }
+        return referencedReports
+    }
+
+    private fun updateReferencedReports(referencedReports: JSONObject, dataPoints: List<JSONObject>) {
+        val referencedReportsFileReferences = referencedReports.keys().asSequence().map { key ->
+            val nestedObject = referencedReports.getJSONObject(key)
+            Pair(nestedObject.getString("fileReference"), nestedObject)
+        }.toMap()
+
+        val dataPointsFileReferences = dataPoints.associate { dataPoint ->
+            val fileReference = dataPoint.getString("fileReference")
+            val fileName = dataPoint.getString("fileName")
+            Pair(fileReference, fileName)
+        } // note that in the case of the same key with two names the last key remains
+
+        val allFileReferences = referencedReportsFileReferences.keys.union(dataPointsFileReferences.keys)
+
+        for (fileReference in allFileReferences) {
+            val referencedReport = referencedReportsFileReferences[fileReference]
+            val dataPointFileName = dataPointsFileReferences[fileReference]
+            updateFileReference(fileReference, referencedReport, dataPointFileName, referencedReports)
+        }
+    }
+
+    private fun updateFileReference(
```

Comment:
There is just one edge case here which is when we have a report in the referencedReports and also in the datapoint (same filereference) but they have different names. In your current function this is skipped. In other words, the report in the referencedReports list maintains its original name and doesn't get the name of the datapoint report. If that is the desired outcome, i'm fine with it.

Reply:
We discussed this in the team and decided that we do not want to deal with inconsistent filenames. So this is not checked. However, we could log if this occurs?

Reply:
its fine the way it is then